AlphaWallet/Common/Services/SpamTokenService.swift

```diff
@@ -0,0 +1,48 @@
+//
+//  SpamTokenService.swift
+//  AlphaWallet
+//
+//  Created by Jerome Chan on 3/4/23.
+//
+/// This service subscribes to the AlphaWalletTokensService for add tokens notifications.
+/// When a spam token is detected, this service will hide it.
+
+import Foundation
+import AlphaWalletFoundation
+import Combine
+
+class SpamTokenService {
+
+    private let tokenGroupIdentifier: TokenGroupIdentifierProtocol
+    private var cancellableSet: Set<AnyCancellable> = []
+    private var tokensService: TokensService
+
+    init(tokenGroupIdentifier: TokenGroupIdentifierProtocol, tokensService: TokensService) {
+        self.tokenGroupIdentifier = tokenGroupIdentifier
+        self.tokensService = tokensService
+    }
+
+    func startMonitoring() {
+        stopMonitoring()
+        tokensService.addedTokensPublisher
+            .receive(on: DispatchQueue.main)
+            .sink { tokens in
+                tokens.forEach { token in
+                    if self.tokenGroupIdentifier.isSpam(address: token.contractAddress.eip55String, chainID: token.server.chainID) {
```

Comment:
Seems the abstraction is done better if instead of:

```
func startMonitoring() {
    stopMonitoring()
    tokensService.addedTokensPublisher
        .receive(on: DispatchQueue.main)
        .sink { tokens in
            tokens.forEach { token in
                if self.tokenGroupIdentifier.isSpam(address: token.contractAddress.eip55String, chainID: token.server.chainID) {
                    self.performSpamHandler(token: token)
                }
            }
        }.store(in: &cancellableSet)
}
```

We do:

```
func startMonitoring() {
    stopMonitoring()
    tokensService.addedTokensPublisher
        .receive(on: DispatchQueue.main)
        .sink { tokens in
            self.performSpamHandler(tokens: tokens)
        }.store(in: &cancellableSet)
```

Reply:
What about
```swift
func startMonitoring() {
    stopMonitoring()
    tokensService.addedTokensPublisher
        .receive(on: DispatchQueue.main)
        .sink { tokens in
            self.performSpamHandler(tokens: tokens.filter{ self.tokenGroupIdentifier.isSpam(address $0.contractAddress.eip55String, chainID: $0.server.chainID} )
        }.store(in: &cancellableSet)
```

Reply:
This one's better because the level of abstraction in `startMonitoring()` is more consistent:

```
self.performSpamHandler(tokens: tokens)
```

Sure, renaming sounds good.